sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/SyntaxTrackers/CSharpBuilderPatternCondition.cs

```diff
@@ -0,0 +1,96 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2020 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+
+namespace SonarAnalyzer.Helpers
+{
+    public class CSharpBuilderPatternCondition : BuilderPatternCondition<InvocationExpressionSyntax>
+    {
+
+        public CSharpBuilderPatternCondition(bool constructorIsSafe, params BuilderPatternDescriptor<InvocationExpressionSyntax>[] descriptors) : base(constructorIsSafe, descriptors) { }
+
+        protected override SyntaxNode RemoveParentheses(SyntaxNode node) =>
+            node.RemoveParentheses();
+
+        protected override SyntaxNode TopMostContainingMethod(SyntaxNode node) =>
+            node.GetTopMostContainingMethod();
+
+        protected override SyntaxNode InvocationExpression(InvocationExpressionSyntax node, out string identifierName)
+        {
+            identifierName = node.Expression.GetIdentifier()?.Identifier.ValueText;
+            return node.Expression;
+        }
+
+        protected override bool IsMemberAccess(SyntaxNode node, out SyntaxNode expression)
```

Comment:
Can be renamed to `TryGetMemberAccessExpression`. Using a try prefix is a common naming pattern for this kind of methods which try to get a result and return a boolean value.

Also the out parameter name, expression, is confusing since both `MemberAccessExpressionSyntax` and  `MemberAccessExpressionSyntax` are expressions and as a caller you don't know, without looking at the implementation, what you get.

Reply:
I've renamed the argument to `memberAccessExpression`. 
I agree with the TryGet naming convention in general, but primary branching condition of these 5 functions is to check the type. No matter what is extracted as secondary output value. TryGet should return false for null, but these functions does not do/check/care about that.

I see that this is not 100% perfect solution, but since we don't have that C# vs. VB.NET layer yet, I think this is good for what we have now.